File: lib/oli/content/writers/writer.ex
defmodule Oli.Content.Writers.Writer do
  alias Oli.Content.Writers.Context

  @type next :: (() -> String.t())
  @type children :: [%{}]

  @callback text(%Context{}, %{}) :: [any()]
  @callback p(%Context{}, next, %{}) :: [any()]
  @callback h1(%Context{}, next, %{}) :: [any()]
  @callback h2(%Context{}, next, %{}) :: [any()]
  @callback h3(%Context{}, next, %{}) :: [any()]
  @callback h4(%Context{}, next, %{}) :: [any()]
  @callback h5(%Context{}, next, %{}) :: [any()]
  @callback h6(%Context{}, next, %{}) :: [any()]
  @callback img(%Context{}, next, %{}) :: [any()]
  @callback youtube(%Context{}, next, %{}) :: [any()]
  @callback audio(%Context{}, next, %{}) :: [any()]
  @callback table(%Context{}, next, %{}) :: [any()]
  @callback tr(%Context{}, next, %{}) :: [any()]
  @callback th(%Context{}, next, %{}) :: [any()]
  @callback td(%Context{}, next, %{}) :: [any()]
  @callback ol(%Context{}, next, %{}) :: [any()]
  @callback ul(%Context{}, next, %{}) :: [any()]
  @callback li(%Context{}, next, %{}) :: [any()]
  @callback math(%Context{}, next, %{}) :: [any()]
  @callback math_line(%Context{}, next, %{}) :: [any()]
  @callback code(%Context{}, next, %{}) :: [any()]
  @callback code_line(%Context{}, next, %{}) :: [any()]
  @callback blockquote(%Context{}, next, %{}) :: [any()]
  @callback a(%Context{}, next, %{}) :: [any()]

  def render(%Context{} = context, content_list, impl) when is_list(content_list) do
    Enum.map(content_list, fn content -> render(context, content, impl) end)
  end

  def render(%Context{} = context, %{"type" => "content", "children" => children}, impl) do
    Enum.map(children, fn child -> render(context, child, impl) end)
  end

  @spec render(Oli.Content.Writers.Context.t(), maybe_improper_list | map, any) :: any
  def render(%Context{} = context, %{"text" => _text} = text_entity, impl) do
    impl.text(context, text_entity)
  end

  def render(%Context{} = context, children, impl) when is_list(children) do
    Enum.map(children, fn child -> render(context, child, impl) end)
  end

  def render(%Context{} = context, %{"type" => type, "children" => children} = entity, impl) do
    next = fn -> render(context, children, impl) end

    case type do
      "p" -> impl.p(context, next, entity)
      "h1" -> impl.h1(context, next, entity)
      "h2" -> impl.h2(context, next, entity)
      "h3" -> impl.h3(context, next, entity)
      "h4" -> impl.h4(context, next, entity)
      "h5" -> impl.h5(context, next, entity)
      "h6" -> impl.h6(context, next, entity)
      "img" -> impl.img(context, next, entity)
      "youtube" -> impl.youtube(context, next, entity)
      "audio" -> impl.audio(context, next, entity)
      "table" -> impl.table(context, next, entity)
      "tr" -> impl.tr(context, next, entity)
      "th" -> impl.th(context, next, entity)
      "td" -> impl.td(context, next, entity)
      "ol" -> impl.ol(context, next, entity)
      "ul" -> impl.ul(context, next, entity)
      "li" -> impl.li(context, next, entity)
      "math" -> impl.math(context, next, entity)
      "math_line" -> impl.math_line(context, next, entity)
      "code" -> impl.code(context, next, entity)
      "code_line" -> impl.code_line(context, next, entity)
      "blockquote" -> impl.blockquote(context, next, entity)
      "a" -> impl.a(context, next, entity)
    end
  end

end

File: assets/src/data/content/writers/writer.ts
import { WriterContext } from './context';
import { ModelElement } from '../model';
import { Text } from 'slate';

export type Next = () => string;
type ElementWriter = (ctx: WriterContext, next: Next, text: ModelElement) => string;

export interface WriterImpl {
  text: (ctx: WriterContext, text: Text) => string;
  p: ElementWriter;
  h1: ElementWriter;
  h2: ElementWriter;
  h3: ElementWriter;
  h4: ElementWriter;
  h5: ElementWriter;
  h6: ElementWriter;
  img: ElementWriter;
  youtube: ElementWriter;
  iframe: ElementWriter;
  audio: ElementWriter;
  table: ElementWriter;
  tr: ElementWriter;
  th: ElementWriter;
  td: ElementWriter;
  ol: ElementWriter;
  ul: ElementWriter;
  li: ElementWriter;
  math: ElementWriter;
  mathLine: ElementWriter;
  code: ElementWriter;
  codeLine: ElementWriter;
  blockquote: ElementWriter;
  a: ElementWriter;
  unsupported: (ctx: WriterContext, element: ModelElement) => string;
}

type ContentItem = { type: 'content'; children: ModelElement[] };
function isContentItem(value: any): value is ContentItem {
  return value && value.type === 'content' && value.children !== undefined;
}

type ContentTypes = ContentItem[] | ContentItem | ModelElement[] | ModelElement | Text;

export class ContentWriter {
  render(context: WriterContext, content: ContentItem[], impl: WriterImpl): string;
  render(context: WriterContext, content: ContentItem, impl: WriterImpl): string;
  render(context: WriterContext, content: ModelElement[], impl: WriterImpl): string;
  render(context: WriterContext, content: ModelElement, impl: WriterImpl): string;
  render(context: WriterContext, content: Text, impl: WriterImpl): string;
  render(context: WriterContext, content: ContentTypes, impl: WriterImpl): string {
    if (Array.isArray(content)) {
      // Typescript seems not to be able to recognize the overloaded function signatures here
      return (content as any).map((item: any) => this.render(context, item, impl)).join('');
    }

    if (isContentItem(content)) {
      return content.children.map((child) => this.render(context, child, impl)).join('');
    }

    if (Text.isText(content)) {
      return impl.text(context, content);
    }

    const next = () => this.render(context, content.children as ModelElement[], impl);

    switch (content.type) {
      case 'p':
        return impl.p(context, next, content);
      case 'h1':
        return impl.h1(context, next, content);
      case 'h2':
        return impl.h2(context, next, content);
      case 'h3':
        return impl.h3(context, next, content);
      case 'h4':
        return impl.h4(context, next, content);
      case 'h5':
        return impl.h5(context, next, content);
      case 'h6':
        return impl.h6(context, next, content);
      case 'img':
        return impl.img(context, next, content);
      case 'youtube':
        return impl.youtube(context, next, content);
      case 'iframe':
        return impl.iframe(context, next, content);
      case 'audio':
        return impl.audio(context, next, content);
      case 'table':
        return impl.table(context, next, content);
      case 'tr':
        return impl.tr(context, next, content);
      case 'th':
        return impl.th(context, next, content);
      case 'td':
        return impl.td(context, next, content);
      case 'ol':
        return impl.ol(context, next, content);
      case 'ul':
        return impl.ul(context, next, content);
      case 'li':
        return impl.li(context, next, content);
      case 'math':
        return impl.math(context, next, content);
      case 'math_line':
        return impl.mathLine(context, next, content);
      case 'code':
        return impl.code(context, next, content);
      case 'code_line':
        return impl.codeLine(context, next, content);
      case 'blockquote':
        return impl.blockquote(context, next, content);
      case 'a':
        return impl.a(context, next, content);
      default:
        return impl.unsupported(context, content);
    }
  }
}
